### DISCUSSION TITLE
What's the difference between `DbType.Time` and `DbType.DateTime`?

### DISCUSSION THREAD

**Discussion by tacosontitan** [dotnet/runtime#103914] [Q&A] [answered]
I'm trying to mark a property with a custom attribute which takes a `DbType` to specify what the database type is for the property, and in this case my property is a `TimeSpan?`:

```csharp
[SqlColumn(nameof(Sample), DbType.Time)]
public TimeSpan? Sample { get; set; }
```

According to the [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.data.dbtype?view=net-8.0) for `DbType`, this is the wrong member to mark with:

> A type representing a SQL Server DateTime value. If you want to use a SQL Server time value, use [Time](https://learn.microsoft.com/en-us/dotnet/api/system.data.sqldbtype?view=net-8.0#system-data-sqldbtype-time).

I can't follow the recommended path of using `SqlDbType.Time` because it's not interchangeable with `DbType`. Furthermore, `DbType` defines a member named `DateTime` which causes further confusion:

> A type representing a date and time value.

---
What's the difference between `DbType.Time` and `DbType.DateTime`?

**Comment by deeprobin** [comment]
@tacosontitan 
The type 'TimeOnly' would be the most suitable naming convention. `TimeSpan` is the period between two points in time.

However, the type `TimeOnly` did not exist at the time the API `DbType.Time` was implemented.
According to the documentation, `DbType.Time` represents a DateTime.

Not even the corefx repo shows me how this was caused. 
This seems to come from the early days of .NET Framework.

@sammonort @ajcvickers @roji As you're the area owners of `System.Data`: Do you know anything about this? If not, could you search for it in your internal .NET Framework repo and maybe find something?

**Comment by huoyaoyuan** [comment, marked as answer]
`DbType` is a maps to ODBC type:

https://github.com/dotnet/runtime/blob/97f50455530cfbe21d28544d2b3fb14611f4ae7a/src/libraries/System.Data.Odbc/src/System/Data/Odbc/Odbc32.cs#L864-L866
https://github.com/dotnet/runtime/blob/97f50455530cfbe21d28544d2b3fb14611f4ae7a/src/libraries/System.Data.Odbc/src/System/Data/Odbc/Odbc32.cs#L770-L772

The official MSSQL documentation mentions mapping for `SQL_TYPE_DATE`(`OdbcType.Date`) and `SQL_TYPE_TIMESTAMP`(`OdbcType.DateTime`), but not `SQL_TYPE.TYPE_TIME`(`OdbcType.Time`). The MSSQL Time type is mapped with `SQL_SS_TIME2`.

According to https://learn.microsoft.com/en-us/sql/relational-databases/native-client/features/date-and-time-improvements?view=sql-server-ver15#use-time-as-a-distinct-data-type , `SQL_TYPE_TIME` maps to `DBTYPE_DBTIME`. Another resource I can find is https://github.com/nanodbc/nanodbc/issues/18 , which suggests `SQL_TYPE_TIME` mapping with MSSQL `time(0)`.

Anyway, this involves a lot of legacies with ODBC and is probably beyond .NET. I'd suggest to not trust the doc and test for what is actually mapped to the DB.
